blog/_posts/2016-03-27-arrays-iteration.md

```diff
@@ -0,0 +1,425 @@
+face value, there are vast armies of programmers out there eagerly
+poised to adopt julia, but who won't even try it because of this
+difference in indexing.  Since recruiting those armies will lead to
+world domination, this is clearly a problem of the utmost urgency.
+
+More seriously, there *are* algorithms which simplify if you can index
+outside of the range from `1:size(A,d)`.  In my own lab's internal
+code, we've long been using a `CenterIndexedArray` type, in which such
+arrays (all of which have odd sizes) are indexed over the range `-n:n`
+and for which 0 refers to the "center" element. One package which
+generalizes this notion is `OffsetArrays`.  Unfortunately, in practice
+both of these array types produce segfaults (due to built-in
+assumptions about when `@inbounds` is appropriate) for many of julia's
+core functions; over time my lab has had to write implementations
+specialized for `CenterIndexedArrays` for quite a few julia functions.
+
+`OffsetArrays` illustrates another conceptual challenge, which can
+easily be demonstrated by `copy!`.  When `dest` is a 1-dimensional
+`OffsetArray` and `src` is a standard `Vector`, what should `copy!`
+do? In particular, where does `src[1]` go? Does it go in the `first`
+element of `dest`, or does it get stored in `dest[1]` (which may not
+be the first element).
+
+Such examples force us to think a little more deeply about what an
+array really is.  There seem to be two potential conceptions.  One is
+that arrays are lists, and multidimensional arrays are
+lists-of-lists-of-lists-of...  In such a world view, the right thing
+to do is to put `src[1]` into the first slot of `dest`, because 1 is
+just a synonym for `first`.  However, this world view doesn't really
+endow any kind of "meaning" to the index-tuple of an array, and in
+that sense doesn't even include the distinction conveyed by an
+`OffsetArray`. In other words, in this world an `OffsetArray` is
+simply nonsensical, and shouldn't exist.
+
+If instead one thinks `OffsetArray`s should exist, this essentially
+forces one to adopt a different world view: arrays are effectively
+associative containers, where each index-tuple is the "key" by which
+one retrieves a value.  With this mode of thinking, `src[1]` should be
+stored in `dest[1]`.
+
+# Formalizing AbstractArray
+
+These examples suggest a formalization of `AbstractArray`:
+
+- `AbstractArray`s are specialized associative containers, in that the
+  allowable "keys" may be restricted by more than just their julia
+  type.  Specifically, the allowable keys must be representable as a
+  cartesian product of one-dimensional lists of values.  The allowed
+  keys may depend not just on the array type but also the specific
+  array (e.g., its size).  Attempted access by keys that cannot be
+  converted to one of the allowed keys, for that specific array,
+  result in `BoundsError`s.
+
+- For any given array, at least one complete independent set of keys
```

Comment:
The "complete set" could also here be defined not by an actual list of values, but rather of intervals (just as today); an  `d`th index `i` is in the set if `1 <= i <= size(itp,d)`. This isn't conceptually much different from how a range object behaves: rather than storing all the values for all the allowed keys, there is an exact definition of which keys are allowed, and the values for those keys might be computed at indexing-time. (Interpolation arrays, like ranges, are also read-only.)

Arguably, though, iteration over interpolation arrays doesn't make much sense. The least weird thing to do (and what Interpolations.jl does today, to my knowledge) is to just return the values on each provided data point. In other words, iterating over an interpolation array yields the same results as iterating over the underlying data set (except that you can't set the value).